src/union.ts

```diff
@@ -179,19 +179,20 @@ export class Union {
             }
 
             // Replace `callback` with our intermediate function.
-            args[lastarg] = (err, resArg: string[] | Buffer[]) => {
-                if(result.size === 0 && err) {
+            args[lastarg] = (err, resArg: string[] | Buffer[] | Dirent[]) => {
+                if(Object.keys(found).length === 0 && err) {
                     return iterate(i + 1, err);
                 }
                 if(resArg) {
-                    // Convert all results to Strings to make sure that they're deduped
                     for (const res of resArg) {
-                        result.add(String(res));
+                        found[res instanceof Buffer ? String(res) : typeof res === 'string' ? res : res.name] = res;
```

Comment:
Would it be possible to break this out into a small arrow function?

Just since it's used in three places, and is an nested ternary; something like `getResultName` would be fine.